### ISSUE TITLE
Unstable adaptive temperature?

### ISSUE THREAD

**Issue by wangh09** [thuml/Transolver#13]
Hi, thanks for the innovative work! I'm trying to reproduce Transover++ with Transover, and I found that the adaptive temperature seems unstable. So I'm wondering if there's any other trick other than a linear layer described in the paper that needs to be applied?

In particular, Eq(3) may involve the following code adding to PhysicsAttention:
In __init__():
`
self.temperature0 = nn.Parameter(torch.ones([1, heads, 1, 1]) * 0.5)
self.to_delta_temp = nn.Linear(head_dim, 1)
nn.init.zeros_(self.to_delta_tau.weight)
nn.init.zeros_(self.to_delta_tau.bias)
`
In forward():
`
temperature = self.temperature0 + self.to_delta_temp(x_mid)
slice_weights = self.softmax(self.in_project_slice(x_mid) / temperature)
`
With above code, gradients may explode and It seems hard to prevent zeros in the temperature variable above. Is there any trick on this? Thanks!

**Comment by luohk19** [comment]
Hi, to stabilize the training process, it's recommended to clamp the values with a minimum threshold (e.g., 0.01, depending on your precision) and apply clip_grad_norm at the beginning of training.

**Comment by wangh09** [comment]
> Hi, to stabilize the training process, it's recommended to clamp the values with a minimum threshold (e.g., 0.01, depending on your precision) and apply clip_grad_norm at the beginning of training.

Thanks for your reply! So basically you do a relu activation after the Linear layer right? or imposing a hard clamp on the whole temperature with 0.01? 

BTW, after some hands-on experience I didn't see a clue that Transolver is superior to a plain Transformer on a small mesh size (~1k) dataset, will Transolver++ be different or that Transolver line-of-work mainly works for larger mesh size? What do you think?
